keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,308 @@
+
+
+### Test Plan
+
+(**FIXME:** Mention that we will add unit tests)
+
+> **Note:** *Not required until targeted at a release.*
+>
+> Consider the following in developing a test plan for this enhancement:
+> - Will there be e2e and integration tests, in addition to unit tests?
+> - How will it be tested in isolation vs with other components?
+>
+> No need to outline all of the test cases, just the general strategy.  Anything
+> that would count as tricky in the implementation and anything particularly
+> challenging to test should be called out.
+>
+> All code is expected to have adequate tests (eventually with coverage
+> expectations).  Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+> when drafting this test plan.
+>
+> [testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+
+### Graduation Criteria
+
+(**FIXME:** Maybe have two major releases of Kustomize for alpha -> beta, write
+user-facing documentation, etc. See
+[here](https://github.com/arrikto/kubernetes-enhancements/blob/master/keps/sig-cli/20200115-kubectl-diff.md#graduation-criteria)
+for an example)
+
+> **Note:** *Not required until targeted at a release.*
+>
+> Define graduation milestones.
+>
+> These may be defined in terms of API maturity, or as something else. The KEP
+> should keep this high-level with a focus on what signals will be looked at to
+> determine graduation.
+>
+> Consider the following in developing the graduation criteria for this enhancement:
+> - [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+> - [Deprecation policy][deprecation-policy]
+>
+> Clearly define what graduation means by either linking to the [API doc
+> definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning),
+> or by redefining what graduation means.
+>
+> In general, we try to use the same stages (alpha, beta, GA), regardless how the
+> functionality is accessed.
+>
+> [maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+> [deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+## Implementation History
+
+(**FIXME:** This section should be left empty, but not be removed)
+
+> Major milestones in the life cycle of a KEP should be tracked in this section.
+> Major milestones might include
+> - the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+> - the `Proposal` section being merged signaling agreement on a proposed design
+> - the date implementation started
+> - the first Kubernetes release where an initial version of the KEP was available
+> - the version of Kubernetes where the KEP graduated to general availability
+> - when the KEP was retired or superseded
+
+## Drawbacks
+
+(**FIXME:** Unclear when to use components vs overlays, anything else?)
+
+> Why should this KEP _not_ be implemented?
+
+## Alternatives
```

Comment:
This is what I propose for the **Alternatives** section:

Since the creation of the kubernetes-sigs/kustomize#1251 issue, various alternatives have been proposed to help in this situation. We're listing them here in order of appearance, along with some comments on why they weren't eventually considered:

* https://github.com/kubernetes-sigs/kustomize/pull/1217: This PR could be used to merge resources with the same GVKN produced by sibling kustomizations. The drawback is that merging YAMLs is not the same as patching the same YAML sequentially. For example, if two kustomizations change different fields of the same resource, it's possible that an old field may resurface, depending on the order of the merges.

* Support sharing patches between kustomizations: Instead of having components and composing overlays from them, we could define patches in a common directory and make overlays mix-and-match them. There were two problems with this approach:

  1. It would require using the `--load_restrictor none` flag of `kustomize build`.
  2. It doesn't cover the case of more complex logic that is easily defined in kustomizations, such as generating and transforming resources.

* Define `generators` / `transformers` and reuse them in overlays: The main problem with this approach is that it requires from the user to redefine the handy generators/transformers of `kustomization.yaml` into separate files, which adds a bit more boilerplate.

* https://github.com/kubernetes-sigs/kustomize/issues/1292: This issue proposed a way to define the behavior of Kustomize when it detects an GVKN collision. This suggestion had the same issue as https://github.com/kubernetes-sigs/kustomize/pull/1217, i.e., that merging YAMLs is not the same as sequentially patching them.

Reply:
>depending on the order of the merges

I think it is not only the order that matters, but also the type of merge/patch. Different merge/patch strategies have specific rules for specific field types, e.g., objects, lists, strings, etc.

Reply:
>Support sharing patches between kustomizations: Instead of having components and composing overlays from them, we could define patches in a common directory and make overlays mix-and-match them.

Actually this description seems to fit to a good extend to our approach, other than the "common directory". I think we need a definite answer about reusability of components, i.e., how multiple top-level overlays refer to the same reusable components. In the workspace file structure we provide above, components live in their own directory, that however is at the same depth with both `base/` and `overlays/`

Reply:
> I think it is not only the order that matters, but also the type of merge/patch. Different merge/patch strategies have specific rules for specific field types, e.g., objects, lists, strings, etc.

I'm referring to the merging of the produced YAMLs here, since the type of the patch is lost when the overlays were merged in this PR.

> Actually this description seems to fit to a good extent to our approach, other than the "common directory". I think we need a definite answer about reusability of components, i.e., how multiple top-level overlays refer to the same reusable components.

I hoped that the "Proposal" section covered this. Do you see any ambiguity there?

> 
> In the workspace file structure we provide above, components live in their own directory, that however is at the same depth with both base/ and overlays/

Yes, the alternative of shared patches resembles that of components. That's why I listed the caveats above, to show some of their limitations.